source/common/http/filter/cors_filter.cc

```diff
@@ -0,0 +1,140 @@
+    : config_(config), is_cors_request_(false) {}
+
+CorsFilter::~CorsFilter() {}
+
+FilterHeadersStatus CorsFilter::decodeHeaders(HeaderMap& headers, bool) {
+  initialize();
+
+  origin_ = headers.Origin();
+  if (origin_ == nullptr || origin_->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  is_cors_request_ = true;
+
+  const Http::HeaderEntry* method = headers.Method();
+  if (method != nullptr && method->value().c_str() != Http::Headers::get().MethodValues.Options) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!cors_enabled_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_origin_ == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  HeaderMapPtr response_headers{
+      new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}}};
+
+  if (allow_credentials_) {
+    response_headers->insertAccessControlAllowCredentials().value(
+        Http::Headers::get().CORSValues.True);
+    response_headers->insertAccessControlAllowOrigin().value(*origin_);
+  } else {
+    response_headers->insertAccessControlAllowOrigin().value(allow_origin_);
+  }
+
+  auto requestMethod = headers.AccessControlRequestMethod();
+  if (requestMethod == nullptr || requestMethod->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_methods_ != "") {
+    response_headers->insertAccessControlAllowMethods().value(allow_methods_);
+  }
+
+  if (allow_headers_ != "") {
+    response_headers->insertAccessControlAllowHeaders().value(allow_headers_);
+  }
+
+  if (expose_headers_ != "") {
+    response_headers->insertAccessControlExposeHeaders().value(expose_headers_);
+  }
+
+  if (max_age_ != "") {
+    response_headers->insertAccessControlMaxAge().value(max_age_);
+  }
+
+  response_headers->insertStatus().value(enumToInt(Http::Code::OK));
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus CorsFilter::encodeHeaders(HeaderMap& headers, bool) {
+  if (!cors_enabled_ || !is_cors_request_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_credentials_) {
+    headers.insertAccessControlAllowCredentials().value(Http::Headers::get().CORSValues.True);
+    headers.insertAccessControlAllowOrigin().value(*origin_);
+  } else {
+    headers.insertAccessControlAllowOrigin().value(allow_origin_);
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+void CorsFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+};
+
+void CorsFilter::initialize() {
+  auto routeEntry = decoder_callbacks_->route()->routeEntry();
+  const Envoy::Router::CorsPolicy& routeCorsPolicy = routeEntry->corsPolicy();
+  const Envoy::Router::CorsPolicy& virtualHostCorsPolicy = routeEntry->virtualHost().corsPolicy();
+
+  if (routeCorsPolicy.enabled() && virtualHostCorsPolicy.enabled()) {
+    cors_enabled_ = true;
```

Comment:
If we opt to stick with the code below, any reason to not early return if cors is not enabled?

Reply:
No reason.